### ISSUE TITLE
AKS POD connecting outside with node src IP address

### ISSUE THREAD

**Issue by kacurez** [Azure/azure-container-networking#460]
**Is this a request for help?**:
 Yes

---
**Is this an ISSUE or FEATURE REQUEST?** (choose one):
Issue

---

**Which release version?**:
Latest

---

**Which component (CNI/IPAM/CNM/CNS)**:
CNI

---

**Which Operating System (Linux/Windows)**:
Linux

---

**For Linux: Include Distro and kernel version using "uname -a"**
Linux aks-agentpool-14016995-vmss000001 4.15.0-1061-azure #66-Ubuntu SMP Thu Oct 3 02:00:50 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

---

**Which Orchestrator and version (e.g. Kubernetes, Docker)**
Kubernetes 1.15.5

---

**What happened**:
We have aks deployed with Azure CNI network plugin(and calico network policy plugin) into a vnet subnet(10.243.0.0/16) and a virtual machine(aka `vm-server`) deployed into the same vnet but a different subnet(10.100.0.0/16). The `vm-server` listens on port 80 and returns the clients IP address when requested with simple http get request. If we connect to the vm-server directly from a POD we noticed the connection comes from the aks node INTERNAL-IP address instead of the PODs IP address itself:

```bash
~ kubectl get nodes -o wide
NAME                                STATUS   ROLES   AGE   VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE             KERNEL-VERSION      CONTAINER-RUNTIME
aks-agentpool-14016995-vmss000000   Ready    agent   30m   v1.15.5   10.243.0.4    <none>        Ubuntu 16.04.6 LTS   4.15.0-1061-azure   docker://3.0.7
aks-agentpool-14016995-vmss000001   Ready    agent   30m   v1.15.5   10.243.0.35   <none>        Ubuntu 16.04.6 LTS   4.15.0-1061-azure   docker://3.0.7
```
Note that the Internal-IP address of one of the nodes is 10.243.0.35.
Now we create a POD spawning a shell inside and curl to vm-server(10.100.0.4) from it :
```
~ kubectl run tmp-shell --generator=run-pod/v1 --image curlimages/curl --rm -it sh
/ $ ifconfig
eth0      Link encap:Ethernet  HWaddr CA:B4:44:67:EB:43  
          inet addr:10.243.0.41  Bcast:0.0.0.0  Mask:255.255.0.0
          UP BROADCAST RUNNING  MTU:1500  Metric:1
          RX packets:5 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:446 (446.0 B)  TX bytes:0 (0.0 B)

/ $ curl 10.100.0.4
{"Clients IP": "10.243.0.35"}
/ $ 
```
The IP address of the POD is 10.243.0.41, however the vm-server "says" the client connected to it with IP 10.243.0.35 and that is node internal IP.
Moreover if we listen to the traffic on the vm-server we can see it received request from the node internal IP address instead of the POD assigned IP address as well:
```
root@MyUbuntuVM:/home/keboola# tcpdump -nei eth0 "(tcp port 80) and (net 10.243)"
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on eth0, link-type EN10MB (Ethernet), capture size 262144 bytes
23:03:50.692764 00:78:88:10:fa:cb > 00:0d:3a:89:83:fa, ethertype IPv4 (0x0800), length 74: 10.243.0.35.11065 > 10.100.0.4.80: Flags [S], seq 2521874544, win 29200, options [mss 1418,sackOK,TS val 3656142044 ecr 0,nop,wscale 7], length 0
23:03:50.692837 00:0d:3a:89:83:fa > 12:34:56:78:9a:bc, ethertype IPv4 (0x0800), length 74: 10.100.0.4.80 > 10.243.0.35.11065: Flags [S.], seq 222204872, ack 2521874545, win 65160, options [mss 1460,sackOK,TS val 3577465302 ecr 3656142044,nop,wscale 7], length 0
```

---

**What you expected to happen**:
The aks POD connects outside of the aks subnet with its own assigned IP address(10.243.0.41 in this case) and not node internal IP address.

---

**How to reproduce it** (as minimally and precisely as possible):
1. Deploy aks cluster with azure CNI networking
2. Deploy a vm machine with ssh to the same vnet as aks cluster but different subnet
3. Ssh to the vm machine and listen for the incoming traffic e.g `tcpdump -nei eth0 "tcp port 1234`
4. run pod inside aks cluster and try to connect from the pod to the deployed vm machine e.g. `curl 10.100.0.4:1234`
5. Note the source IP address 

The whole resources and scripts to create such environment can be found here:
https://github.com/keboola/azure-kubernetes-sandbox/tree/master#aks-pod-network-problem

The same problem can be also simulated if we deny all incoming traffic on the vm-server subnet from node internal ip address(deny src 10.243.0.35 in this case) and try to connect to the vm-server from POD - the connection request will timeout.

---

**Anything else we need to know**:
We run 3rd party docker containers from inside aks PODs and we need this to isolate the running 3rd party docker container from the rest of our platform(internal db..). The whole problem is described in more details here:
https://gist.github.com/kacurez/adb8a606464ca9cbe19d14c293c4407f
Similar approach works in AWS ECS, where we assign an IP address to an ECS task via [vpc networking mode](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-networking.html).

---

**Comment by tamilmani1989** [comment]
@kacurez It looks like traffic is getting snatted. Can you send us output of following command.?
`sudo iptables -t nat -L POSTROUTING`

Also what is your vnet address space?

**Comment by kacurez** [comment]
@tamilmani1989 vnet space address is 10.0.0.0/8 and basically there are 2 subnets, one for the aks cluster(10.243.0.0/16) and the other one for the vm-server(10.100.0.0/16)
Here is the output of `sudo iptables -t nat -L` command from one of the node instancies with IP 10.243.0.15:
<details>
<summary> sudo iptables -t nat -L </summary>

```
root@aks-agentpool-14016995-vmss000001:/# sudo iptables -t nat -L 
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         
cali-PREROUTING  all  --  anywhere             anywhere             /* cali:6gwbT8clXdHdC1b1 */
KUBE-SERVICES  all  --  anywhere             anywhere             /* kubernetes service portals */
DOCKER     all  --  anywhere             anywhere             ADDRTYPE match dst-type LOCAL

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
cali-OUTPUT  all  --  anywhere             anywhere             /* cali:tVnHkvAo15HuiPy0 */
KUBE-SERVICES  all  --  anywhere             anywhere             /* kubernetes service portals */
DOCKER     all  --  anywhere            !127.0.0.0/8          ADDRTYPE match dst-type LOCAL

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
cali-POSTROUTING  all  --  anywhere             anywhere             /* cali:O3lYWMrLQYEMJtB5 */
KUBE-POSTROUTING  all  --  anywhere             anywhere             /* kubernetes postrouting rules */
MASQUERADE  all  --  172.17.0.0/16        anywhere            

Chain DOCKER (2 references)
target     prot opt source               destination         
RETURN     all  --  anywhere             anywhere            

Chain KUBE-MARK-DROP (0 references)
target     prot opt source               destination         
MARK       all  --  anywhere             anywhere             MARK or 0x8000

Chain KUBE-MARK-MASQ (14 references)
target     prot opt source               destination         
MARK       all  --  anywhere             anywhere             MARK or 0x4000

Chain KUBE-NODEPORTS (1 references)
target     prot opt source               destination         

Chain KUBE-POSTROUTING (1 references)
target     prot opt source               destination         
MASQUERADE  all  --  anywhere             anywhere             /* kubernetes service traffic requiring SNAT */ mark match 0x4000/0x4000

Chain KUBE-SEP-4ZW5ZOAQIRC2HPF3 (1 references)
target     prot opt source               destination         
KUBE-MARK-MASQ  all  --  10.243.0.63          anywhere            
DNAT       tcp  --  anywhere             anywhere             tcp to:10.243.0.63:9090

Chain KUBE-SEP-CGZFNS4C4QEXG2ZL (1 references)
target     prot opt source               destination         
KUBE-MARK-MASQ  all  --  10.243.0.38          anywhere            
DNAT       udp  --  anywhere             anywhere             udp to:10.243.0.38:53

Chain KUBE-SEP-CQ2EHHSAI6KY7EH6 (1 references)
target     prot opt source               destination         
KUBE-MARK-MASQ  all  --  10.243.0.58          anywhere            
DNAT       tcp  --  anywhere             anywhere             tcp to:10.243.0.58:443

Chain KUBE-SEP-NMZDJ3X7MTBO5PD4 (1 references)
target     prot opt source               destination         
KUBE-MARK-MASQ  all  --  10.243.0.38          anywhere            
DNAT       tcp  --  anywhere             anywhere             tcp to:10.243.0.38:53

Chain KUBE-SEP-QSJAC3DLSZZ6R7XQ (1 references)
target     prot opt source               destination         
KUBE-MARK-MASQ  all  --  10.243.0.35          anywhere            
DNAT       tcp  --  anywhere             anywhere             tcp to:10.243.0.35:5473

Chain KUBE-SEP-R67KCS2QK6ZUPCBZ (1 references)
target     prot opt source               destination         
KUBE-MARK-MASQ  all  --  20.40.136.11         anywhere            
DNAT       tcp  --  anywhere             anywhere             tcp to:20.40.136.11:443

Chain KUBE-SEP-WMWL3D623YT57BCY (1 references)
target     prot opt source               destination         
KUBE-MARK-MASQ  all  --  10.243.0.30          anywhere            
DNAT       udp  --  anywhere             anywhere             udp to:10.243.0.30:53

Chain KUBE-SEP-ZR4Z6WTLYFR6QIRT (1 references)
target     prot opt source               destination         
KUBE-MARK-MASQ  all  --  10.243.0.30          anywhere            
DNAT       tcp  --  anywhere             anywhere             tcp to:10.243.0.30:53

Chain KUBE-SERVICES (2 references)
target     prot opt source               destination         
KUBE-MARK-MASQ  udp  -- !10.243.0.0/16        10.1.0.10            /* kube-system/kube-dns:dns cluster IP */ udp dpt:domain
KUBE-SVC-TCOU7JCQXEZGVUNU  udp  --  anywhere             10.1.0.10            /* kube-system/kube-dns:dns cluster IP */ udp dpt:domain
KUBE-MARK-MASQ  tcp  -- !10.243.0.0/16        10.1.0.10            /* kube-system/kube-dns:dns-tcp cluster IP */ tcp dpt:domain
KUBE-SVC-ERIFXISQEP7F7OF4  tcp  --  anywhere             10.1.0.10            /* kube-system/kube-dns:dns-tcp cluster IP */ tcp dpt:domain
KUBE-MARK-MASQ  tcp  -- !10.243.0.0/16        10.1.172.7           /* kube-system/kubernetes-dashboard: cluster IP */ tcp dpt:http
KUBE-SVC-XGLOHA7QRQ3V22RZ  tcp  --  anywhere             10.1.172.7           /* kube-system/kubernetes-dashboard: cluster IP */ tcp dpt:http
KUBE-MARK-MASQ  tcp  -- !10.243.0.0/16        10.1.195.14          /* kube-system/metrics-server: cluster IP */ tcp dpt:https
KUBE-SVC-LC5QY66VUV2HJ6WZ  tcp  --  anywhere             10.1.195.14          /* kube-system/metrics-server: cluster IP */ tcp dpt:https
KUBE-MARK-MASQ  tcp  -- !10.243.0.0/16        10.1.0.1             /* default/kubernetes:https cluster IP */ tcp dpt:https
KUBE-SVC-NPX46M4PTMTKRN6Y  tcp  --  anywhere             10.1.0.1             /* default/kubernetes:https cluster IP */ tcp dpt:https
KUBE-MARK-MASQ  tcp  -- !10.243.0.0/16        10.1.227.96          /* kube-system/calico-typha:calico-typha cluster IP */ tcp dpt:5473
KUBE-SVC-VUMZCXJHDF5H4EWS  tcp  --  anywhere             10.1.227.96          /* kube-system/calico-typha:calico-typha cluster IP */ tcp dpt:5473
KUBE-NODEPORTS  all  --  anywhere             anywhere             /* kubernetes service nodeports; NOTE: this must be the last rule in this chain */ ADDRTYPE match dst-type LOCAL

Chain KUBE-SVC-ERIFXISQEP7F7OF4 (1 references)
target     prot opt source               destination         
KUBE-SEP-ZR4Z6WTLYFR6QIRT  all  --  anywhere             anywhere             statistic mode random probability 0.50000000000
KUBE-SEP-NMZDJ3X7MTBO5PD4  all  --  anywhere             anywhere            

Chain KUBE-SVC-LC5QY66VUV2HJ6WZ (1 references)
target     prot opt source               destination         
KUBE-SEP-CQ2EHHSAI6KY7EH6  all  --  anywhere             anywhere            

Chain KUBE-SVC-NPX46M4PTMTKRN6Y (1 references)
target     prot opt source               destination         
KUBE-SEP-R67KCS2QK6ZUPCBZ  all  --  anywhere             anywhere            

Chain KUBE-SVC-TCOU7JCQXEZGVUNU (1 references)
target     prot opt source               destination         
KUBE-SEP-WMWL3D623YT57BCY  all  --  anywhere             anywhere             statistic mode random probability 0.50000000000
KUBE-SEP-CGZFNS4C4QEXG2ZL  all  --  anywhere             anywhere            

Chain KUBE-SVC-VUMZCXJHDF5H4EWS (1 references)
target     prot opt source               destination         
KUBE-SEP-QSJAC3DLSZZ6R7XQ  all  --  anywhere             anywhere            

Chain KUBE-SVC-XGLOHA7QRQ3V22RZ (1 references)
target     prot opt source               destination         
KUBE-SEP-4ZW5ZOAQIRC2HPF3  all  --  anywhere             anywhere            

Chain cali-OUTPUT (1 references)
target     prot opt source               destination         
cali-fip-dnat  all  --  anywhere             anywhere             /* cali:GBTAv2p5CwevEyJm */

Chain cali-POSTROUTING (1 references)
target     prot opt source               destination         
cali-fip-snat  all  --  anywhere             anywhere             /* cali:Z-c7XtVd2Bq7s_hA */
cali-nat-outgoing  all  --  anywhere             anywhere             /* cali:nYKhEzDlr11Jccal */

Chain cali-PREROUTING (1 references)
target     prot opt source               destination         
cali-fip-dnat  all  --  anywhere             anywhere             /* cali:r6XmIziWUJsdOK6Z */

Chain cali-fip-dnat (2 references)
target     prot opt source               destination         

Chain cali-fip-snat (1 references)
target     prot opt source               destination         

Chain cali-nat-outgoing (1 references)
target     prot opt source               destination         
MASQUERADE  all  --  anywhere             anywhere             /* cali:flqWnvo8yq4ULQLa */ match-set cali40masq-ipam-pools src ! match-set cali40all-ipam-pools dst
root@aks-agentpool-14016995-vmss000001:/# 

```
</details>

I am not sure if I read it correctly but looks like traffic is getting snatted as you stated. If that's true and it is also an expected behaviour then can we do something to make sure the POD is connecting with its own IP address to anywhere within the vnet?
Basically we need to distinguish the connection from POD and from the 3rd party docker container run inside the POD. The only way we see it is possible is via source IP where POD is using its own assigned IP address and the 3rd party container with mounted docker socket from host is using node internal IP address.

**Comment by tamilmani1989** [comment]
@kacurez I dont see ip masq agent rules in your VM. Whenever you create AKS/AKS-Engine cluster, you find ip-masq-agent pod running in all nodes.  These are kuber system daemonsets running in my cluster and you can see **azure-ip-masq-agent**. Can you check yours? 

```
rnmadmin@k8s-master-34648761-0:~$ kubectl get daemonsets -n kube-system
NAME                         DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR                 AGE
azure-cni-networkmonitor     3         3         3       3            3           beta.kubernetes.io/os=linux   104d
azure-ip-masq-agent          3         3         3       3            3           beta.kubernetes.io/os=linux   104d
blobfuse-flexvol-installer   2         2         1       2            1           beta.kubernetes.io/os=linux   104d
keyvault-flexvolume          2         2         1       2            1           beta.kubernetes.io/os=linux   104d
kube-proxy                   3         3         3       3            3           beta.kubernetes.io/os=linux   104d
```

**Comment by kacurez** [comment]
@tamilmani1989 there is no such azure-ip-masq-agent daemonset running in my cluster, see:

```
~ kubectl get daemonsets -n kube-system
NAME          DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR                 AGE
calico-node   2         2         2       2            2           beta.kubernetes.io/os=linux   9m45s
kube-proxy    2         2         2       2            2           beta.kubernetes.io/os=linux   9m46s
```

**Comment by tamilmani1989** [comment]
@kacurez You have to check with AKS team. They manage ip-masq-agent daemonset deployment. Create a ticket against this repo [https://github.com/Azure/AKS](url) . They could help you with this. Feel free to engage me if you/they need additional information

**Comment by kacurez** [comment]
@tamilmani1989 thanks for tip! Looking at the AKS issues I found one mostly related to this issue:
https://github.com/Azure/AKS/issues/1207
But if I understand it correctly my problem here is missing ip-masq-agent in the cluster so you recommend to report an issue about ip-masq-agent daemonset not being deployed into aks cluster when it should be?

**Comment by tamilmani1989** [comment]
@kacurez Yes it would be better if you create separate ticket and understand why ip-masq-agent was not running in your setup.

**Comment by kacurez** [comment]
@tamilmani1989 so azure-ip-masq-agent was not deployed because we are deploying aks cluster with calico kubernetes network policies plugin. If we deploy aks with azure network policy(or none) and azure network plugin then azure-ip-masq-agent is deployed as well. With only azure network plugin POD connects to vm server using the IP address originally assigned to the pod and not node as desired.
Thanks for help!
